# Problem 3

Implement a program to find the solution to the 8 queens problem using a depth-first search. Start with an empty board and place one queen at a time in consecutive columns and in order.
The provided code should reflect the pseudo-code presented in class, starting from the initial state `[∅, ∅, ∅, ∅, ∅, ∅, ∅, ∅]` and reaching any of the possible goal states (e.g., `[5, 7, 4, 1, 3, 8, 6, 2]` shown in `Figure 3`).

<p align="center">
  <img align="center" width="400" alt="figure1" src="./images/figure3.png"/>
</p>


# Main Objectives

-   Implement the 8 queens problem using a depth-first search.
-   Solve the problem using the provided pseudo-code.
-   Implement the pseudo-code in python


# Implementation

The pseudo-code to implement the problem is shown below:

```pseudocode
clase Nodo:
    función __init__(self, estado):
        self.estado = estado
        self.padre = Nulo

    función __str__(self) -> cadena:
        resultado = "Estado actual (cadena): {}\n".format(self.estado)
        resultado += "Tablero:\n"
        resultado += self.estado_a_tablero(self.estado) + "\n"
        retornar resultado

    función __repr__(self) -> cadena:
        self.__str__()

    función estado_a_tablero(estado):
        tablero = ""
        para fila en estado:
            tablero += ' '.join('Q' si str(i) == fila sino '.' para i en rango(8)) + "\n"
        retornar tablero.strip()

clase ProblemaOchoReinas:
    función __init__(self, estadoInicial):
        self.nodoInicial = Nodo(estadoInicial)

    función esMeta(self, nodo):
        retornar '∅' no está en nodo.estado

    función acciones(self, nodo):
        si '∅' no está en nodo.estado:
            retornar []
        col = nodo.estado.index('∅')
        retornar [str(fila) para fila en rango(8) si no self.conflicto(nodo.estado, col, fila)]

    función transición(self, nodo, acción):
        nueva_lista_estado = lista(nodo.estado)
        col = nueva_lista_estado.index('∅')
        nueva_lista_estado[col] = acción
        retornar Nodo(''.join(nueva_lista_estado))

    función expandir(self, nodo):
        retornar [self.transición(nodo, acción) para acción en self.acciones(nodo)]

    función conflicto(self, estado, col, fila):
        para c, f en enumerar(estado):
            si f != '∅' y (int(f) == fila o abs(int(f) - fila) == abs(c - col)):
                retornar Verdadero
        retornar Falso

función BúsquedaDFS(problema):
    frontera = [problema.nodoInicial]
    i = 0
    mientras frontera:
        nodo = frontera.pop()
        imprimir(nodo)
        si problema.esMeta(nodo):
            imprimir("Iteraciones:", i)
            retornar nodo
        frontera.extend(invertido(problema.expandir(nodo)))
        i += 1
    retornar Nulo

estadoInicial = "∅∅∅∅∅∅∅∅"
nodoSolucion = BúsquedaDFS(ProblemaOchoReinas(estadoInicial))
si nodoSolucion:
    imprimir("Se encontró la solución final:")
    imprimir(nodoSolucion)
sino:
    imprimir("No se encontró una solución.")

```


### Class `Node`:

-   The `Node` class represents a node in the search space. Each node has a state (represented as a string) and a link to its parent node.
-   The `__str__` method returns a readable representation of the node, showing its state and corresponding chessboard.
-   The `state_to_board` method converts the node's state into a visual representation of the chessboard.

### Class `EightQueensProblem`:

-   The `EightQueensProblem` class encapsulates the eight queens problem. It takes an initial state representing the empty chessboard.
-   The `isGoal` method checks if a given node represents a goal state where no queens are attacking each other.
-   The `actions` method returns the possible actions that can be taken from a given node, i.e., placing a queen in an empty column.
-   The `transition` method performs a given action on a node, placing a queen in an empty column.
-   The `expand` method generates all possible successor nodes from a given node.

### Function `DFS`:

-   The `DFS` function implements the depth-first search (DFS) algorithm. It starts with an initial node and explores all successor nodes deeply before backtracking.
-   It uses a stack (the `frontier` variable) to store nodes that have not yet been explored.
-   While there are nodes in the frontier, it pops a node from the stack, prints it, and checks if it's a solution. If it's not, it expands the node and adds the successors to the stack.
-   It returns the solution node if found, otherwise returns `None`.

### Main Flow:

-   An empty initial state is defined for the eight queens problem.
-   DFS is executed on the eight queens problem using the initial state.
-   If a solution is found, the solution node is printed; otherwise, a message indicating that no solution was found is printed.


# Implementación in Python

The implementation of the 8 queens problem using depth-first search in Python is shown below:


In [10]:
class Node:
    def __init__(self, state):
        self.state = state
        self.parent = None

    def __str__(self) -> str:
        result = "Estado actual (string): {}\n".format(self.state)
        result += "Tablero:\n"
        result += self.state_to_board(self.state) + "\n"
        return result
    
    def __repr__(self) -> str:
        self.__str__()

    @staticmethod
    def state_to_board(state):
        board = ""
        for row in state:
            board += ' '.join('Q' if str(i) == row else '.' for i in range(8)) + "\n"
        return board.strip()

class EightQueensProblem:
    def __init__(self, initialState):
        self.initialNode = Node(initialState)

    def isGoal(self, node):
        return '∅' not in node.state

    def actions(self, node):
        if '∅' not in node.state:
            return []
        col = node.state.index('∅')
        return [str(row) for row in range(8) if not self.conflict(node.state, col, row)]

    def transition(self, node, action):
        new_state_list = list(node.state)
        col = new_state_list.index('∅')
        new_state_list[col] = action
        return Node(''.join(new_state_list))

    def expand(self, node):
        return [self.transition(node, action) for action in self.actions(node)]

    def conflict(self, state, col, row):
        for c, r in enumerate(state):
            if r != '∅' and (int(r) == row or abs(int(r) - row) == abs(c - col)):
                return True
        return False

def DFS(problem):
    frontier = [problem.initialNode]
    i = 0
    while frontier:
        node = frontier.pop()
        # print(node) # To see state of the game

        if problem.isGoal(node):
            print("Iteraciones:", i)
            return node
        frontier.extend(reversed(problem.expand(node)))
        i += 1
    return None


# Results

To test the implementation, we will run the code with the initial state `∅∅∅∅∅∅∅∅` and check if a solution is found.


In [11]:
initialState = "∅∅∅∅∅∅∅∅"
solutionNode = DFS(EightQueensProblem(initialState))

if solutionNode:
    print("Se encontró la solución final:")
    print(solutionNode)
else:
    print("No se encontró una solución.")

Iteraciones: 113
Se encontró la solución final:
Estado actual (string): 04752613
Tablero:
Q . . . . . . .
. . . . Q . . .
. . . . . . . Q
. . . . . Q . .
. . Q . . . . .
. . . . . . Q .
. Q . . . . . .
. . . Q . . . .



The provided solution is:

```txt
Iteraciones: 113
Se encontró la solución final:
Estado actual (string): 04752613
Tablero:
Q . . . . . . .
. . . . Q . . .
. . . . . . . Q
. . . . . Q . .
. . Q . . . . .
. . . . . . Q .
. Q . . . . . .
. . . Q . . . .
```


We can print all the steps to reach the solution by uncommenting the line `print(nodo)` in the `DFS` function.
But it took 113 iterations to find the solution. So we will only print the final solution.

Our solution `04752613` is correct and the code is working as expected.


# Conclusion

The 8 queens problem is a classic problem in computer science that involves placing eight queens on an 8x8 chessboard such that no two queens attack each other. We implemented a solution to this problem using depth-first search in Python, following the provided pseudo-code. The code successfully finds a solution (`04752613`) to the problem by exploring the search space and backtracking when necessary. The final solution shows a valid placement of queens on the chessboard where no two queens are attacking each other. The depth-first search algorithm is an effective way to solve the 8 queens problem and can be used to find solutions to other combinatorial problems as well.
